cmake/doc/doxygen.cmake

```diff
@@ -0,0 +1,40 @@
+find_package(Doxygen)
+
+if (DOXYGEN_FOUND)
+
+    # Set doxygen settings here
+    set(DOXYGEN_PROJECT_NAME HiveMind)
+    set(DOXYGEN_OUTPUT_DIRECTORY doc)
+    set(DOXYGEN_COLLABORATION_GRAPH YES)
+    set(DOXYGEN_EXTRACT_ALL YES)
+    set(DOXYGEN_CLASS_DIAGRAMS YES)
+    set(DOXYGEN_HIDE_UNDOC_RELATIONS NO)
+    set(DOXYGEN_HAVE_DOT YES)
+    set(DOXYGEN_CLASS_GRAPH YES)
+    set(DOXYGEN_CALL_GRAPH YES)
+    set(DOXYGEN_CALLER_GRAPH YES)
+    set(DOXYGEN_COLLABORATION_GRAPH YES)
+    set(DOXYGEN_BUILTIN_STL_SUPPORT YES)
+    set(DOXYGEN_EXTRACT_PRIVATE YES)
+    set(DOXYGEN_EXTRACT_PACKAGE YES)
+    set(DOXYGEN_EXTRACT_STATIC YES)
+    set(DOXYGEN_EXTRACT_LOCALMETHODS YES)
+    set(DOXYGEN_UML_LOOK YES)
+    set(DOXYGEN_UML_LIMIT_NUM_FIELDS 50)
+    set(DOXYGEN_TEMPLATE_RELATIONS YES)
+    set(DOXYGEN_DOT_GRAPH_MAX_NODES 100)
+    set(DOXYGEN_MAX_DOT_GRAPH_DEPTH 0)
+    set(DOXYGEN_DOT_TRANSPARENT YES)
+
+    if(ENABLE_WARNINGS_AS_ERROR )
+        set(DOXYGEN_WARN_AS_ERROR YES)
+    endif()
+
+    doxygen_add_docs(doc
+        ALL
+        ${PROJECT_SOURCE_DIR}/src
```

Comment:
Do you think we should maybe remove the tests from the documentation? I have a feeling it will end up cluttering the documentation.

Reply:
I'll add them. Should cmake hard fail if doxygen and clang-tidy is missing or simply throw an error message? Obviously I would add a flag to disable the hard fail if needed, but it would be on by default.

Reply:
I'm guessing you're commenting on the .readme suggestion?
If so, I think just a warning as we currently have is enough, it's not like if we absolutely need to generate the doc or check clang-tidy when building